### PR TITLE
Changes to GroupRequest logic, AtmosphereProcess interfaces, and AD handling of initial conditions.

### PR DESCRIPTION
**Pull request by bartgol**
I apologize for the size of the PR, but the more I tried to accommodate something, the more I realized it was by far easier if I modified something else.

This PR started with one goal: fix how Homme creates the Q_prev copy of the tracers group. It ended up doing more. Here's a tentative list of what it does:

- reworked implementation of `contiguous_superset`. I think the number of LOC increased, but it is imho easier to follow now.
- changed names/impl of GroupRequests (GR) derived on other GRs. E.g., Relationship->DerivationType, Child->Subset, Parent->Superset.
- make FM correctly handle copy of groups: Copy only allowed on the same grid, and the individual fields are not created to avoid clashes with the subfields of original group.
- move most impl of AtmosphereProcess (AP) class into a cpp file (the hpp was getting too long, making hard to see the class interfaces at a glance).
- store all fields/groups in/out in the base AP class, provide getters for derived class (so they can't add more fields to the list, which could mess up some infrastructure logic, like that there is 1-1 correspondence between m_fields_in and the required fields requests).
- allow derived classes to not override any set_blah_impl method, unless *really* needed. All basic bookkeping is already in the AP class
- make AD load IC for field on any grid. This does not yet support loading fields from nc file directly on dyn grid. But at least one can init a dyn grid field to a constant or equal to another field on the phys grid (via syntax `field_1: field_2, Physics_GLL`).

There might be more, but it's late, and I need to go to bed. I wanted to put this in before taking off for the long weekend.

A couple of things that I did not address:
- Homme needs Q_prev only for ftype 0, but currently it always creates it. Fix this by introducing some checks on ftype.
- I realized we have some pb when creating a subgroup of a group, asking for it to be bundled. E.g., SHOC wants a subset of the tracers, and would like them bundled. However, we never create the subfield `Q_shoc=Q(:,1:num_shoc_tr,:)`, since we can only handle subfield that completely slice away a dimension (rather than taking an extent). I have an idea in mind on how to extend this, to allow, say, `subfield(const int idim, const int k0, const int n)`, with `k0` being the 1st index along dimension `idim` to subview, and `n` being the number of slices. But it may take some time, so we need to see how big of a priority it is (I suspect we want it in the long run though).

### CODE REVIEW

File: components/scream/src/share/field/field_group.hpp
Comment (jeff-cohere) [inline components/scream/src/share/field/field_group.hpp:134]: Doesn't this predicate mean that `EKAT_REQUIRE_MSG` will report the error below if `m_bundle` is set? If so, what does it mean that we're checking `m_bundle` on line 137?

File: components/scream/src/share/atm_process/atmosphere_process.hpp
Comment (tcclevenger) [inline components/scream/src/share/atm_process/atmosphere_process.hpp:122]: If you have other changes to make, here's a typo "haave".

File: components/scream/src/share/atm_process/atmosphere_process.hpp
Comment (tcclevenger) [inline components/scream/src/share/atm_process/atmosphere_process.hpp:305]: Typo: "methods"

File: components/scream/src/share/field/field_request.hpp
Comment (tcclevenger) [inline components/scream/src/share/field/field_request.hpp:5]: Maybe this is a feature I don't know, but should there be 2 slashes here?

**Review by tcclevenger — COMMENTED** [review]
This looks great, makes the code much nicer. I've looked at most of the files and have a few comments (mostly just typos, so feel free to ignore if it needs no changes before merging). I'll take a look at the rest later today or Monday if it still isn't merged.

File: components/scream/src/share/field/field_group.hpp
Comment (bartgol) [reply components/scream/src/share/field/field_group.hpp:134]: Uhm, it reports the error if neither of the conditions pass. The grid name is pulled from any of the fields in the group, or from the bundled field. If neither are set, we can't retrieve the grid name. Since a group may have `m_fields` set, but not the bundle, or viceversa, at 137 we check which one is set, and grab the grid name from the first one that is available.

File: components/scream/src/share/field/field_request.hpp
Comment (bartgol) [reply components/scream/src/share/field/field_request.hpp:5]: I've seen that same typo happening multiple times on my PRs. It might be the autocomplete and me having a race condition for adding the slash, I am not sure... But no, it's not a feature. :)

File: components/scream/tests/uncoupled/rrtmgp/input.yaml
Comment (jeff-cohere) [inline components/scream/tests/uncoupled/rrtmgp/input.yaml:21]: Seems like this comment might no longer be accurate.

File: components/scream/tests/uncoupled/homme/input.yaml
Comment (jeff-cohere) [inline components/scream/tests/uncoupled/homme/input.yaml:8]: What about just `File`? It's already in the `Initial Conditions` section/block

Comment (bartgol) [reply components/scream/tests/uncoupled/homme/input.yaml:8]: I agree.

File: components/scream/src/share/field/field_group.hpp
Comment (jeff-cohere) [reply components/scream/src/share/field/field_group.hpp:134]: Got it. Thanks!

File: components/scream/src/dynamics/homme/atmosphere_dynamics.cpp
Comment (jeff-cohere) [inline components/scream/src/dynamics/homme/atmosphere_dynamics.cpp:167]: Not a huge deal, but `pass` -> `passes`

File: components/scream/src/dynamics/homme/atmosphere_dynamics.cpp
Comment (jeff-cohere) [inline components/scream/src/dynamics/homme/atmosphere_dynamics.cpp:185]: Any reason to keep this comment?

**Review by jeff-cohere — APPROVED** [review]
Nice work, Luca. This PR does clear things up quite a bit.

Since we're introducing and clarifying more concepts (groups, importing, etc), it would be good if we wrote down a description of these concepts and their interaction somewhere (Confluence, maybe, for the moment?). This stuff needs to go into whatever we put together for a "SCREAM developer guide". I know Peter C has mentioned the need for documentation in recent days. Maybe we're getting to the point where we need to think about it more seriously.

File: components/scream/src/dynamics/homme/atmosphere_dynamics.cpp
Comment (bartgol) [reply components/scream/src/dynamics/homme/atmosphere_dynamics.cpp:185]: Nope.

**Comment by bartgol** [comment]
@jeff-cohere I agree. I will make this my next todo list item. Once the PR is in (and the homme forcing fix PR in e3sm is up), I will carve time to write down some documentation. Pinky promise. ;)

**Comment by jeff-cohere** [comment]
> @jeff-cohere I agree. I will make this my next todo list item. Once the PR is in (and the homme forcing fix PR in e3sm is up), I will carve time to write down some documentation. Pinky promise. ;)

Let me know how I can help. You don't have to document all of this yourself!

**Comment by bartgol** [comment]
> > @jeff-cohere I agree. I will make this my next todo list item. Once the PR is in (and the homme forcing fix PR in e3sm is up), I will carve time to write down some documentation. Pinky promise. ;)
> 
> Let me know how I can help. You don't have to document all of this yourself!

When I start, I will ping you, and we can see if we can thread the execution.

**Review by jeff-cohere — APPROVED** [review]
Looks good!

**Review by tcclevenger — APPROVED** [review]
Nice work!

File: components/scream/src/control/tests/ad_tests.yaml
Comment (AaronDonahue) [inline components/scream/src/control/tests/ad_tests.yaml:11]: Not relevant to this PR, but I'm mentioning now because it just reminded me.  The YAML parser interprets any number starting with `-` as a string not a double.  So it's impossible to assign a negative value in this way.  We may want to fix that at some point.

Comment (jeff-cohere) [reply components/scream/src/control/tests/ad_tests.yaml:11]: Do you mean that if you tell the parser to interpret negative numbers as doubles, that operation fails? The parser can interpret anything as a string, and you tell it to cast to a specific data type, right?

Comment (jeff-cohere) [reply components/scream/src/control/tests/ad_tests.yaml:11]: Or are we using something on top of yaml-cpp to parse our YAML files?

File: components/scream/src/physics/p3/atmosphere_microphysics.cpp
Comment (AaronDonahue) [inline components/scream/src/physics/p3/atmosphere_microphysics.cpp:186]: Did we mean to keep these commented out?

File: components/scream/src/physics/shoc/atmosphere_macrophysics.cpp
Comment (AaronDonahue) [inline components/scream/src/physics/shoc/atmosphere_macrophysics.cpp:350]: Same question here, did we mean to keep these commented out?

File: components/scream/src/control/tests/ad_tests.yaml
Comment (bartgol) [reply components/scream/src/control/tests/ad_tests.yaml:11]: Really?!? Wow, I didn't know that. There should be a way to fix this.

File: components/scream/src/physics/p3/atmosphere_microphysics.cpp
Comment (bartgol) [reply components/scream/src/physics/p3/atmosphere_microphysics.cpp:186]: I guess I can remove them. I wasn't sure what to do with the FPC's in Homme, so I commented for now. You are working on fixing these checks, right? Does it bother you to keep them in master just for a couple more days, till you get your PR ready?

File: components/scream/src/control/tests/ad_tests.yaml
Comment (bartgol) [reply components/scream/src/control/tests/ad_tests.yaml:11]: Interesting. In ekat, the unit test for yaml parser correctly reads stuff like `My Int: -2` as a negative int. But `-2.0` is read as a string. I guess I should fix this right away in ekat.

File: components/scream/src/share/tests/field_tests.cpp
Comment (AaronDonahue) [inline components/scream/src/share/tests/field_tests.cpp:542]: Do we mean to keep this commented line?

Comment (bartgol) [reply components/scream/src/share/tests/field_tests.cpp:542]: Ah yeah. This line is simply explaining how group3 is "extended" by the following group request. Maybe I can improve the comment.

**Comment by AaronDonahue** [comment]
Looks good @bartgol .  As for the property checks, I think we can revisit the ones set up in each individual process.  I envisioned that individual atm. processes may have had some strict criteria for incoming or outgoing fields so we would expect to see those property checks added in each process.  Also, for simple testing we can't rely on another process to add a property check.  Consider for example the negative tracers property, if we are running SHOC stand-alone we would still want to catch those.  I think we should approve this for now and add a v1.1 task to figure out how we want to give individual processes the power to check fields of interest.

**Review by AaronDonahue — APPROVED** [review]

**Comment by bartgol** [comment]
> Looks good @bartgol . As for the property checks, I think we can revisit the ones set up in each individual process. I envisioned that individual atm. processes may have had some strict criteria for incoming or outgoing fields so we would expect to see those property checks added in each process. Also, for simple testing we can't rely on another process to add a property check. Consider for example the negative tracers property, if we are running SHOC stand-alone we would still want to catch those. I think we should approve this for now and add a v1.1 task to figure out how we want to give individual processes the power to check fields of interest.

I agree. Each process should store FPC's _internally_, and run them before/after run. The FPC's stored in the field should only be used if we foresee the need to run said check before/after all (or at least many) atm procs.

**Pull request merged**
